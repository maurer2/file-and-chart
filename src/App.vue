<template>
  <div id="app">
    <h1 class="title">File-and-chart</h1>

    <section class="section">
      <h2 class="subtitle">File input</h2>
      <Form @fileUpload='handleFormChange' />
    </section>

    <section class="section" v-if="showTable">
      <h2 class="subtitle">Data table</h2>
      <Table :tableData="inputData" />
    </section>

    <section class="section" v-if="showCharts">
      <h2 class="subtitle">Charts</h2>
      <Charts :chartData="inputData" />
    </section>
  </div>
</template>

<script>
import { csvParse } from 'd3';

import Form from './components/Form.vue';
import Table from './components/Table.vue';
import Charts from './components/Charts.vue';

export default {
  name: 'app',
  data() {
    return {
      inputData: [],
    }
  },
  components: {
    Form,
    Table,
    Charts,
  },
  computed: {
    showTable() {
      return this.inputData.length > 0;
    },
    showCharts() {
      return this.inputData.length > 0;
    }
  },
  methods: {
    handleFormChange(inputData) {
      const inputParsed = csvParse(inputData);

      this.inputData = inputParsed;
    }
  },
};

</script>

<style lang="scss">
html {
  box-sizing: border-box;
  font-size: 16px;
}

*,
*:before,
*:after {
  box-sizing: inherit;
}

body {
  margin: 0;
  padding: 1rem;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

</style>

<style lang="scss" scoped>
#app {
  border: 1px solid black;
  text-align: center;
  color: #2c3e50;
  width: 100%;
}

.section {
  padding: 1.5rem 1rem;
  border-top: 1px solid black;
}

.title {
  margin: 2rem 1rem;
}

.subtitle {
  margin-top: 0;
  margin-bottom: 1rem;
}

</style>
