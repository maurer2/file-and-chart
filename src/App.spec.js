import { shallowMount } from '@vue/test-utils';
import App from '@/App.vue';

describe('App', () => {
  let wrapper;

  const mockedHandleFormChange = jest.fn();

  beforeEach(() => {
    wrapper = shallowMount(App, {});
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('App snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('App contains all elements and child elements', () => {
    expect(wrapper.find('.app').exists()).toBe(true);
    expect(wrapper.find('.title').exists()).toBe(true);
    expect(wrapper.findAll('.subtitle').exists()).toBe(true);
  });

  test('App has only form section when inputData is not set', () => {
    expect(wrapper.findAll('.section').length).toBe(1);
    expect(wrapper.find('form-stub').exists()).toBe(true);
    expect(wrapper.find('table-stub').exists()).toBe(false);
    expect(wrapper.find('charts-stub').exists()).toBe(false);
  });

  test('App has table section when inputData is set', () => {
    expect(wrapper.find('table-stub').exists()).toBe(false);

    wrapper.setData({ inputData: ['cat'] });

    expect(wrapper.find('table-stub').exists()).toBe(true);
  });

  test('App has chart section when inputData is set', () => {
    expect(wrapper.find('charts-stub').exists()).toBe(false);

    wrapper.setData({ inputData: ['cat'] });

    expect(wrapper.find('charts-stub').exists()).toBe(true);
  });

  test('Getter showTable returns correct value when inputData is set', () => {
    expect(wrapper.vm.showTable).toBe(false);

    wrapper.setData({ inputData: ['cat'] });

    expect(wrapper.vm.showTable).toBe(true);
  });

  test('Getter showCharts returns correct value when inputData is set', () => {
    expect(wrapper.vm.showCharts).toBe(false);

    wrapper.setData({ inputData: ['cat'] });

    expect(wrapper.vm.showCharts).toBe(true);
  });

  test('Form handleFormChange method is called when fileUpload event is triggered from child', () => {
    wrapper.setMethods({ handleFormChange: mockedHandleFormChange });

    wrapper.find('form-stub').vm.$emit('fileUpload');

    expect(mockedHandleFormChange).toBeCalled();
  });
});
