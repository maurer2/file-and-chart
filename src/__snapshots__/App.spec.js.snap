// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App snapshot 1`] = `
<div id="app" class="app">
  <h1 class="title">
    File-and-chart
  </h1>
  <section class="section">
    <h2 class="subtitle">
      File input
    </h2>
    <form-stub></form-stub>
  </section>
  <!---->
  <!---->
</div>
`;
